File: harness/coppice/internal/web/static/_tests/tiles.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import {
  LOCKED, newTiles, fill, open, closeSlot, swap, rotate, reset, focused, slotOf, shown,
} from '../tiles.js';

// These cases mirror internal/tiles/tiles_test.go one for one. The Go
// model drives the terminal floor and this one drives the phone page, and
// the two must give the same slots for the same clicks.

const count = (slots, pane) => slots.filter((s) => s === pane).length;

test('fill replaces the focused slot only', () => {
  const t = newTiles('focus', 2);
  fill(t, 'p1'); t.focus = 1; fill(t, 'p2'); t.focus = 0;
  fill(t, 'p3');
  assert.deepEqual(t.slots, ['p3', 'p2']);
});

test('a pane has at most one tile', () => {
  const t = newTiles('focus', 2);
  fill(t, 'p1'); t.focus = 1;
  fill(t, 'p1');
  assert.equal(count(t.slots, 'p1'), 1);
});

test('swap keeps focus with the pane', () => {
  const t = newTiles('focus', 2);
  fill(t, 'p1'); t.focus = 1; fill(t, 'p2'); t.focus = 0;
  assert.deepEqual(swap(t, 0, 1), { ok: true });
  assert.equal(focused(t), 'p1');
  assert.equal(t.slots[1], 'p1');
});

test('locked refuses swap', () => {
  const t = newTiles('focus', 2);
  t.locked = true;
  const r = swap(t, 0, 1);
  assert.equal(r.ok, false);
  assert.equal(r.reason, 'Layout is locked. Unlock first.');
  assert.equal(r.reason, LOCKED);
});

test('an empty slot wins over focus', () => {
  const t = newTiles('focus', 2);
  fill(t, 'p1');
  fill(t, 'p2');
  assert.deepEqual(t.slots, ['p1', 'p2']);
});

test('newTiles makes the slots each layout starts with', () => {
  assert.deepEqual(newTiles('focus', 3), { slots: ['', '', ''], focus: 0, layout: 'focus', locked: false });
  assert.equal(newTiles('focus', 0).slots.length, 1);
  assert.deepEqual(newTiles('one', 3).slots, ['']);
  assert.deepEqual(newTiles('all', 3).slots, []);
  assert.equal(newTiles('nonsense', 2).layout, 'focus');
  assert.equal(focused(newTiles('focus', 2)), '');
});

test('fill under all appends and under one replaces', () => {
  const all = newTiles('all', 0);
  fill(all, 'p1'); fill(all, 'p2');
  assert.deepEqual(all.slots, ['p1', 'p2']);
  const one = newTiles('one', 0);
  fill(one, 'p1'); fill(one, 'p2');
  assert.deepEqual(one.slots, ['p2']);
});

test('fill focuses the slot it filled', () => {
  const t = newTiles('focus', 3);
  fill(t, 'p1'); fill(t, 'p2');
  assert.equal(t.focus, 1);
  fill(t, 'p1');
  assert.equal(t.focus, 0);
});

test('fill and open ignore an empty name', () => {
  const t = newTiles('focus', 2);
  fill(t, '');
  fill(t, undefined);
  assert.deepEqual(t.slots, ['', '']);
  assert.equal(open(t, ''), -1);
  assert.deepEqual(t.slots, ['', '']);
});

test('open appends and focuses a new slot', () => {
  const t = newTiles('focus', 2);
  fill(t, 'p1'); fill(t, 'p2');
  assert.equal(open(t, 'p3'), 2);
  assert.equal(t.focus, 2);
  assert.deepEqual(t.slots, ['p1', 'p2', 'p3']);
});

test('open on a pane already in a slot returns that slot', () => {
  const t = newTiles('focus', 2);
  fill(t, 'p1'); fill(t, 'p2');
  assert.equal(open(t, 'p1'), 0);
  assert.equal(t.focus, 0);
  assert.equal(t.slots.length, 2);
});

test('open under one replaces slot zero', () => {
  const t = newTiles('one', 0);
  fill(t, 'p1');
  assert.equal(open(t, 'p2'), 0);
  assert.deepEqual(t.slots, ['p2']);
});

test('closeSlot under all removes the slot and clamps focus', () => {
  const t = newTiles('all', 0);
  fill(t, 'p1'); fill(t, 'p2'); fill(t, 'p3');
  t.focus = 2;
  closeSlot(t, 2);
  assert.deepEqual(t.slots, ['p1', 'p2']);
  assert.equal(t.focus, 1);
  closeSlot(t, 0);
  assert.deepEqual(t.slots, ['p2']);
  assert.equal(t.focus, 0);
  closeSlot(t, 0);
  assert.deepEqual(t.slots, []);
  assert.equal(t.focus, 0);
  assert.equal(focused(t), '');
});

test('closeSlot under focus empties the slot', () => {
  const t = newTiles('focus', 2);
  fill(t, 'p1'); fill(t, 'p2');
  closeSlot(t, 0);
  assert.deepEqual(t.slots, ['', 'p2']);
  closeSlot(t, 5);
  closeSlot(t, -1);
  assert.deepEqual(t.slots, ['', 'p2']);
});

test('swap refuses an index outside the slots', () => {
  const t = newTiles('focus', 2);
  const r = swap(t, 0, 2);
  assert.equal(r.ok, false);
  assert.equal(r.reason, 'Slot 2 is outside the slots 0 to 1.');
  assert.equal(swap(t, -1, 0).ok, false);
  assert.equal(swap(t, 0, 1.5).ok, false);
});

test('locked refuses rotate and reset and changes nothing', () => {
  const t = newTiles('focus', 2);
  fill(t, 'p1');
  t.locked = true;
  assert.deepEqual(rotate(t), { ok: false, reason: LOCKED });
  assert.deepEqual(reset(t, ['p2']), { ok: false, reason: LOCKED });
  assert.deepEqual(t.slots, ['p1', '']);
});

test('rotate moves every slot right and keeps focus with its pane', () => {
  const t = newTiles('focus', 3);
  fill(t, 'p1'); fill(t, 'p2'); fill(t, 'p3');
  t.focus = 1;
  assert.deepEqual(rotate(t), { ok: true });
  assert.deepEqual(t.slots, ['p3', 'p1', 'p2']);
  assert.equal(t.focus, 2);
  assert.equal(focused(t), 'p2');
  assert.deepEqual(rotate(newTiles('all', 0)), { ok: true });
});

test('reset under each layout', () => {
  const order = ['a', 'b', 'c'];
  const all = newTiles('all', 0);
  fill(all, 'z');
  assert.deepEqual(reset(all, order), { ok: true });
  assert.deepEqual(all.slots, ['a', 'b', 'c']);
  assert.equal(all.focus, 0);
  order[0] = 'changed';
  assert.equal(all.slots[0], 'a', 'reset under all shares the caller\'s array');

  const focus = newTiles('focus', 2);
  fill(focus, 'z'); focus.focus = 1;
  reset(focus, ['a', 'b', 'c']);
  assert.deepEqual(focus.slots, ['a', 'b']);
  assert.equal(focus.focus, 0);
  const wide = newTiles('focus', 3);
  fill(wide, 'z');
  reset(wide, ['a']);
  assert.deepEqual(wide.slots, ['a', '', '']);

  const one = newTiles('one', 0);
  fill(one, 'z');
  reset(one, ['a', 'b']);
  assert.deepEqual(one.slots, ['a']);
  assert.equal(one.focus, 0);
  reset(one, []);
  assert.deepEqual(one.slots, ['']);
});

test('slotOf and shown', () => {
  const t = newTiles('focus', 3);
  fill(t, 'p1'); fill(t, 'p2');
  assert.equal(slotOf(t, 'p2'), 1);
  assert.equal(slotOf(t, 'p9'), -1);
  assert.equal(slotOf(t, ''), -1);
  assert.deepEqual(shown(t, 2), ['p1', 'p2']);
  assert.deepEqual(shown(t, 5), ['p1', 'p2', '']);
  assert.deepEqual(shown(t, 0), []);
  assert.deepEqual(shown(t, -1), []);
  const got = shown(t, 1);
  got[0] = 'changed';
  assert.equal(t.slots[0], 'p1', 'shown shares the slots array');
});

File: harness/coppice/internal/tiles/tiles_test.go
package tiles

import (
	"errors"
	"testing"
)

func count(slots []string, pane string) int {
	n := 0
	for _, s := range slots {
		if s == pane {
			n++
		}
	}
	return n
}

func TestFillReplacesTheFocusedSlotOnly(t *testing.T) {
	ts := New("focus", 2)
	ts.Fill("p1")
	ts.Focus = 1
	ts.Fill("p2")
	ts.Focus = 0
	ts.Fill("p3")
	if ts.Slots[0] != "p3" || ts.Slots[1] != "p2" {
		t.Fatalf("slots %v", ts.Slots)
	}
}

func TestAPaneHasAtMostOneTile(t *testing.T) {
	ts := New("focus", 2)
	ts.Fill("p1")
	ts.Focus = 1
	ts.Fill("p1")
	if n := count(ts.Slots, "p1"); n != 1 {
		t.Fatalf("p1 in %d slots", n)
	}
}

func TestSwapKeepsFocusWithThePane(t *testing.T) {
	ts := New("focus", 2)
	ts.Fill("p1")
	ts.Focus = 1
	ts.Fill("p2")
	ts.Focus = 0
	if err := ts.Swap(0, 1); err != nil {
		t.Fatal(err)
	}
	if ts.Focused() != "p1" || ts.Slots[1] != "p1" {
		t.Fatalf("focus did not follow: %v focus=%d", ts.Slots, ts.Focus)
	}
}

func TestLockedRefusesSwap(t *testing.T) {
	ts := New("focus", 2)
	ts.Locked = true
	if err := ts.Swap(0, 1); err == nil {
		t.Fatal("swap while locked")
	}
}

func TestEmptySlotWinsOverFocus(t *testing.T) {
	ts := New("focus", 2)
	ts.Fill("p1")
	ts.Fill("p2")
	if ts.Slots[1] != "p2" || ts.Slots[0] != "p1" {
		t.Fatalf("empty slot not used: %v", ts.Slots)
	}
}

func TestNewMakesTheSlotsEachLayoutStartsWith(t *testing.T) {
	if got := len(New(Focus, 3).Slots); got != 3 {
		t.Fatalf("focus 3: %d slots", got)
	}
	if got := len(New(Focus, 0).Slots); got != 1 {
		t.Fatalf("focus 0 clamps to 1, got %d", got)
	}
	if got := len(New(One, 3).Slots); got != 1 {
		t.Fatalf("one: %d slots", got)
	}
	if got := len(New(All, 3).Slots); got != 0 {
		t.Fatalf("all: %d slots", got)
	}
	if ts := New(Focus, 2); ts.Focused() != "" {
		t.Fatalf("a new tile set has a focused pane: %q", ts.Focused())
	}
}

func TestFillUnderAllAppendsAndUnderOneReplaces(t *testing.T) {
	all := New(All, 0)
	all.Fill("p1")
	all.Fill("p2")
	if len(all.Slots) != 2 || all.Slots[1] != "p2" {
		t.Fatalf("all: %v", all.Slots)
	}
	one := New(One, 0)
	one.Fill("p1")
	one.Fill("p2")
	if len(one.Slots) != 1 || one.Slots[0] != "p2" {
		t.Fatalf("one: %v", one.Slots)
	}
}

func TestFillFocusesTheSlotItFilled(t *testing.T) {
	ts := New(Focus, 3)
	ts.Fill("p1")
	ts.Fill("p2")
	if ts.Focus != 1 {
		t.Fatalf("focus %d after filling slot 1", ts.Focus)
	}
	ts.Fill("p1")
	if ts.Focus != 0 {
		t.Fatalf("focus %d after filling a pane that already has a slot", ts.Focus)
	}
}

func TestFillIgnoresAnEmptyName(t *testing.T) {
	ts := New(Focus, 2)
	ts.Fill("")
	if ts.Slots[0] != "" || ts.Slots[1] != "" {
		t.Fatalf("an empty name changed the slots: %v", ts.Slots)
	}
	if i := ts.Open(""); i != -1 {
		t.Fatalf("Open of an empty name returned %d", i)
	}
}

func TestOpenAppendsAndFocusesANewSlot(t *testing.T) {
	ts := New(Focus, 2)
	ts.Fill("p1")
	ts.Fill("p2")
	if i := ts.Open("p3"); i != 2 || ts.Focus != 2 || len(ts.Slots) != 3 {
		t.Fatalf("open returned %d, slots %v, focus %d", i, ts.Slots, ts.Focus)
	}
}

func TestOpenOnAPaneAlreadyInASlotReturnsThatSlot(t *testing.T) {
	ts := New(Focus, 2)
	ts.Fill("p1")
	ts.Fill("p2")
	if i := ts.Open("p1"); i != 0 || ts.Focus != 0 || len(ts.Slots) != 2 {
		t.Fatalf("open returned %d, slots %v, focus %d", i, ts.Slots, ts.Focus)
	}
}

func TestOpenUnderOneReplacesSlotZero(t *testing.T) {
	ts := New(One, 0)
	ts.Fill("p1")
	if i := ts.Open("p2"); i != 0 || len(ts.Slots) != 1 || ts.Slots[0] != "p2" {
		t.Fatalf("open returned %d, slots %v", i, ts.Slots)
	}
}

func TestCloseSlotUnderAllRemovesTheSlot(t *testing.T) {
	ts := New(All, 0)
	ts.Fill("p1")
	ts.Fill("p2")
	ts.Fill("p3")
	ts.Focus = 2
	ts.CloseSlot(2)
	if len(ts.Slots) != 2 || ts.Focus != 1 {
		t.Fatalf("slots %v focus %d", ts.Slots, ts.Focus)
	}
	ts.CloseSlot(0)
	if len(ts.Slots) != 1 || ts.Slots[0] != "p2" || ts.Focus != 0 {
		t.Fatalf("slots %v focus %d", ts.Slots, ts.Focus)
	}
	ts.CloseSlot(0)
	if len(ts.Slots) != 0 || ts.Focus != 0 || ts.Focused() != "" {
		t.Fatalf("slots %v focus %d", ts.Slots, ts.Focus)
	}
}

func TestCloseSlotUnderFocusEmptiesTheSlot(t *testing.T) {
	ts := New(Focus, 2)
	ts.Fill("p1")
	ts.Fill("p2")
	ts.CloseSlot(0)
	if len(ts.Slots) != 2 || ts.Slots[0] != "" || ts.Slots[1] != "p2" {
		t.Fatalf("slots %v", ts.Slots)
	}
	ts.CloseSlot(5)
	if len(ts.Slots) != 2 {
		t.Fatalf("an index outside the slots changed them: %v", ts.Slots)
	}
}

func TestCloseSlotClampsFocusUnderAll(t *testing.T) {
	ts := New(All, 0)
	ts.Fill("p1")
	ts.Fill("p2")
	ts.Focus = 1
	ts.CloseSlot(1)
	if ts.Focus != 0 || ts.Focused() != "p1" {
		t.Fatalf("focus %d focused %q", ts.Focus, ts.Focused())
	}
}

func TestSwapRefusesAnIndexOutsideTheSlots(t *testing.T) {
	ts := New(Focus, 2)
	err := ts.Swap(0, 2)
	if err == nil {
		t.Fatal("swap with an index outside the slots was accepted")
	}
	if got := err.Error(); got != "Slot 2 is outside the slots 0 to 1." {
		t.Fatalf("error text %q", got)
	}
	if err := ts.Swap(-1, 0); err == nil {
		t.Fatal("swap with a negative index was accepted")
	}
}

func TestLockedRefusesRotateAndReset(t *testing.T) {
	ts := New(Focus, 2)
	ts.Fill("p1")
	ts.Locked = true
	if err := ts.Rotate(); !errors.Is(err, ErrLocked) {
		t.Fatalf("rotate while locked: %v", err)
	}
	if err := ts.Reset([]string{"p2"}); !errors.Is(err, ErrLocked) {
		t.Fatalf("reset while locked: %v", err)
	}
	if err := ts.Swap(0, 1); !errors.Is(err, ErrLocked) {
		t.Fatalf("swap while locked: %v", err)
	}
	if ts.Slots[0] != "p1" {
		t.Fatalf("a refused call changed the slots: %v", ts.Slots)
	}
}

func TestRotateMovesEverySlotRightAndKeepsFocusWithItsPane(t *testing.T) {
	ts := New(Focus, 3)
	ts.Fill("p1")
	ts.Fill("p2")
	ts.Fill("p3")
	ts.Focus = 1
	if err := ts.Rotate(); err != nil {
		t.Fatal(err)
	}
	want := []string{"p3", "p1", "p2"}
	for i := range want {
		if ts.Slots[i] != want[i] {
			t.Fatalf("slots %v", ts.Slots)
		}
	}
	if ts.Focus != 2 || ts.Focused() != "p2" {
		t.Fatalf("focus %d focused %q", ts.Focus, ts.Focused())
	}
	empty := New(All, 0)
	if err := empty.Rotate(); err != nil {
		t.Fatalf("rotate with no slots: %v", err)
	}
}

func TestResetUnderEachLayout(t *testing.T) {
	order := []string{"a", "b", "c"}

	all := New(All, 0)
	all.Fill("z")
	all.Focus = 0
	if err := all.Reset(order); err != nil {
		t.Fatal(err)
	}
	if len(all.Slots) != 3 || all.Slots[2] != "c" || all.Focus != 0 {
		t.Fatalf("all: %v focus %d", all.Slots, all.Focus)
	}
	order[0] = "changed"
	if all.Slots[0] != "a" {
		t.Fatal("reset under all shares the caller's slice")
	}

	focus := New(Focus, 2)
	focus.Fill("z")
	focus.Focus = 1
	if err := focus.Reset([]string{"a", "b", "c"}); err != nil {
		t.Fatal(err)
	}
	if len(focus.Slots) != 2 || focus.Slots[0] != "a" || focus.Slots[1] != "b" || focus.Focus != 0 {
		t.Fatalf("focus: %v focus %d", focus.Slots, focus.Focus)
	}
	wide := New(Focus, 3)
	wide.Fill("z")
	if err := wide.Reset([]string{"a"}); err != nil {
		t.Fatal(err)
	}
	if wide.Slots[0] != "a" || wide.Slots[1] != "" || wide.Slots[2] != "" {
		t.Fatalf("focus with a short order: %v", wide.Slots)
	}

	one := New(One, 0)
	one.Fill("z")
	if err := one.Reset([]string{"a", "b"}); err != nil {
		t.Fatal(err)
	}
	if len(one.Slots) != 1 || one.Slots[0] != "a" || one.Focus != 0 {
		t.Fatalf("one: %v focus %d", one.Slots, one.Focus)
	}
	if err := one.Reset(nil); err != nil {
		t.Fatal(err)
	}
	if one.Slots[0] != "" {
		t.Fatalf("one with an empty order: %v", one.Slots)
	}
}

func TestSlotOfAndShown(t *testing.T) {
	ts := New(Focus, 3)
	ts.Fill("p1")
	ts.Fill("p2")
	if i, ok := ts.SlotOf("p2"); !ok || i != 1 {
		t.Fatalf("SlotOf p2 = %d, %v", i, ok)
	}
	if _, ok := ts.SlotOf("p9"); ok {
		t.Fatal("SlotOf found a pane that has no slot")
	}
	if _, ok := ts.SlotOf(""); ok {
		t.Fatal("SlotOf matched an empty slot to an empty name")
	}
	if got := ts.Shown(2); len(got) != 2 || got[0] != "p1" || got[1] != "p2" {
		t.Fatalf("Shown(2) = %v", got)
	}
	if got := ts.Shown(5); len(got) != 3 {
		t.Fatalf("Shown(5) = %v", got)
	}
	if got := ts.Shown(0); len(got) != 0 {
		t.Fatalf("Shown(0) = %v", got)
	}
	if got := ts.Shown(-1); len(got) != 0 {
		t.Fatalf("Shown(-1) = %v", got)
	}
	got := ts.Shown(1)
	got[0] = "changed"
	if ts.Slots[0] != "p1" {
		t.Fatal("Shown shares the slots slice")
	}
}
